### PR TITLE
Add Create User endpoint

### PR DESCRIPTION
**Pull request by jmthibault79**
DSDEEPB-1769
- [x] **Submitter**: Rebase to develop. DO NOT SQUASH
- [x] **Submitter**: Make sure Swagger is updated if API changes
- [x] **Submitter**: Make sure documentation for code is complete
- [x] **Submitter**: Review code comments; remove done TODOs, create stories for remaining TODOs
- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Add description or comments on the PR explaining the hows/whys (if not obvious)
- [x] Tell  that the PR exists if he wants to look at it
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
- [x] **LR**: Initial review by LR and others.
- [x] Comment / review / update cycle:
  - Rest of team may comments on PR at will
  - **LR assigns to submitter** for feedback fixes
  - Submitter updates documentation as needed
  - Submitter rebases to develop again if necessary
  - Submitter makes further commits. DO NOT SQUASH. **Reassign to LR** for further feedback
- [x]  sign off
- [x] **LR** sign off
- [x] **Assign to submitter** to finalize
- [x] **Submitter**: Squash commits, rebase if necessary
- [x] **Submitter**: Verify all tests go green, including CI tests
- [x] **Submitter**: Merge to develop 
- [x] **Submitter**: Delete branch after merge
- [x] **Submitter**: Check configuration files in Jenkins in case they need changes
- [ ] **Submitter**: Verify swagger UI on dev server still works after deployment
- [x] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/rawls/Boot.scala [hunk 1]
```diff
@@ -52,12 +52,14 @@ object Boot extends App {
       dataSource.shutdown()
     }
 
+    val authDAO = new GraphAuthDAO()
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/Boot.scala]: put this in the GraphContainerDAO and pass that to the auth dao

File: src/main/scala/org/broadinstitute/dsde/rawls/webservice/UserApiService.scala [hunk 2]
```diff
@@ -24,6 +22,15 @@ trait UserApiService extends HttpService with PerRequestCreator with UserInfoDir
   val userServiceConstructor: UserInfo => UserService
 
   val userRoutes = requireUserInfo() { userInfo =>
+    path("user") {
+      post {
+        entity(as[RawlsUser]) { userToCreate =>
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/webservice/UserApiService.scala]: this end point should not take any post body parameters. All information should come from OIDC headers as the post body is not trusted.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphAuthDAO.scala [hunk 3]
```diff
@@ -12,6 +12,15 @@ import scala.collection.JavaConversions._
...
 
+  override def createUser(rawlsUser: RawlsUser, txn: RawlsTransaction): RawlsUser = txn withGraph { db =>
+    getUserVertex(db, rawlsUser) match {
+      case Some(_) => throw new RawlsException("Cannot create user %s in database because it already exists".format(rawlsUser))
+      case None =>
+        saveObject[RawlsUser](rawlsUser, addVertex(db, VertexSchema.User), None, db)
+        rawlsUser
+    }
+  }
+
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphAuthDAO.scala:30]: Why do we want this separate implementation to `saveUser` immediately below?

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphAuthDAO.scala [hunk 4]
```diff
@@ -26,14 +35,14 @@ class GraphAuthDAO extends AuthDAO with GraphDAO {
...
-      case None => throw new RawlsException("Cannot delete group %s from database because it does not exist".format(rawlsGroup.groupName))
+      case None => throw new RawlsException("Cannot delete group %s from database because it does not exist".format(rawlsGroup))
       case Some(vertex) => removeObject(vertex, db)
     }
   }
 
   private def createGroup(rawlsGroup: RawlsGroup, txn: RawlsTransaction): RawlsGroup = txn withGraph { db =>
     getGroupVertex(db, rawlsGroup) match {
-      case Some(_) => throw new RawlsException("Cannot create group %s in database because it already exists".format(rawlsGroup.groupName))
+      case Some(_) => throw new RawlsException("Cannot create group %s in database because it already exists".format(rawlsGroup))
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphAuthDAO.scala:52]: Ditto here.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphAuthDAO.scala (hunk 3 above)
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphAuthDAO.scala:30]: Save works like other Graph saves - meaning either create or update.

Create is for the case where we know the user or group should not be present, and therefore it throws an error if it finds them.

File: src/main/scala/org/broadinstitute/dsde/rawls/webservice/UserApiService.scala (hunk 2 above)
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/webservice/UserApiService.scala]: I think I was confused about whether this was an admin endpoint, and therefore required a different user than the OIDC headers.

But no, that is not the case, so OIDC is fine.  I'll update it to work that way.

File: src/main/scala/org/broadinstitute/dsde/rawls/Boot.scala (hunk 1 above)
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/Boot.scala]: done

**Comment by jmthibault79** [comment]
Comments addressed.

**Comment by dvoet** [comment]
:+1:

**Comment by jmthibault79** [comment]
Travis was successful but it's not updating here for some reason.

**Pull request merged**